text/3266-rust-semver-2.md

```diff
@@ -0,0 +1,446 @@
+
+; SemVer 2
+version = numbers ["-" pre-release] ["+" build]
+
+numbers = num-ident dot num-ident dot num-ident
+
+pre-release = pre-ident *(dot pre-ident)
+pre-ident = alphanum-ident / num-ident
+
+build = build-ident *(dot build-ident)
+build-ident = 1*(ALPHA / DIGIT / ident-join)
+
+; need at least one alpha or ident-join
+alphanum-ident = *DIGIT (ALPHA / ident-join) *(ALPHA / DIGIT / ident-join)
+
+; leading zero are not accepted like 01 or 001
+num-ident = num-ident-non-zero / "0"
+num-ident-non-zero = digit-non-zero *DIGIT
+digit-non-zero = "1" / "2" / "3" / "4" / "5" / "6" / "7" / "8" / "9"
+
+ident-join = "-"
+dot = "."
+```
+
+We thank authors of [`bap`] and [`abnfgen`] to have provided free tool to test this ABNF. [`abnfgen`] can be used to generate tests for parser that want to implement Rust SemVer 2.
+
+### Cargo
+[cargo]: #cargo
+
+`Cargo.toml` has 2 new fields:
+
+* `package.rust-semver`, define what Rust SemVer version Cargo should use for this `Cargo.toml`, the default is `1` for 2021 edition of Rust.
+
+* `dependencies.foo.allow-advanced-operator`, the default is `warn` for `rust-semver = "1"` and `deny` for `rust-semver = "2"`:
+
+  * `warn` will emit a warning if a requirement version use `~` or `||` operator.
+  * `allow` will accept tilde `~` or `||` operator.
+  * `deny` will emit an error if a requirement version use `~` or `||` operator.
+
+`package.version` will now default to `0.0.0`.
+
+Cargo should emit a warning when `^` should be used instead of `~`. It is the case when only the MAJOR and MINOR are specified in `~` like `~1.2` == `^1.2` or `~0.2.5` == `^0.2.5`.
+
+### Pre-release Guideline
+[pre-release-guideline]: #pre-release-guideline
+
+A pre-release tag MAY follow such convention:
+
+* Alpha pre-release are considered very unstable that is similar to `0.0.z`. Alpha version should not have any compatibility expectation. It's recommended for an Alpha pre-release to set version like `1.0.0-alpha.PREMAJOR` where `PREMAJOR` is a numeric identifier incremented at each Alpha pre-release. `1.0.0-alpha.0` and `1.0.0-alpha.1` don't have any compatibility expectation.
+* Beta pre-release are considered unstable that are similar to `0.y.z` when `y > 0`. Beta version may have compatibility expectation. It's recommended for a Beta pre-release to set version like `1.0.0-beta.PREMAJOR.PREMINOR` where `PREMAJOR` is a numeric identifier incremented at each breaking change and `PREMINOR` is a numeric identifier incremented at each non-breaking change. `1.0.0-beta.0.1` MAY be compatible with `1.0.0-beta.0.0` or `1.0.0-beta.2.0` MAY be compatible with `1.0.0-beta.2.4`
+* If a pre-release is considerate to be the last pre-release before a release we call them Candidate Release and if a crate desire hint their users about Candidate Release, it's recommended to use build tag like `1.0.0-beta.4.0+rc`. There is no compatible expectation between a Candidate Release and the final release. They may be any number of Candidate Release.
```

Comment:
imo this complicates things and deviates from pre-existing norms in various ecosystems for not enough gain to warrant it

Reply:
Maybe I wasn't clear but these guideline are not warrant at all. There are just what "could" be a pre-release convention for a crate. This guideline is arbitrary, if you have another proposition I all hear.

The main advantage of current guide that it's "follow" the `0.0.z` <=> `0.alpha.z` and `0.y.z` <=> `beta.y.z` convention of Cargo, thus like you said "all pre-releases of clap v3 was mostly breaking change", this guide would so advice to never use beta, and stay in alpha for clap case since it was very unstable.

Reply:
I'm a bit confused then what the role of this within the RFC is if this is more of a brainstorm for a convention for crates to follow but that cargo will not be respecting it.  I lean towards removing it completely.  If it is kept, the reason it is included needs to be made more clear and it needs to be re-iterated that cargo treats all pre-releases as incompatible with each other and that this is purely for communicating to users.